File: IAIso-v5.0/core/iaiso-node/src/conformance/consent.ts
/**
 * Conformance runner — consent vectors.
 *
 * Validates scope matching, valid token verification, invalid token
 * rejection, and issue-and-verify roundtrips.
 */

import { readFileSync } from "node:fs";
import { join } from "node:path";

import {
  ConsentIssuer,
  ConsentVerifier,
  ExpiredToken,
  InvalidToken,
  RevokedToken,
  scopeGranted,
} from "../consent/index.js";
import type { VectorResult } from "./pressure.js";

interface ConsentVectorFile {
  version: string;
  hs256_key_shared: string;
  scope_match: Array<{
    name: string;
    granted: string[];
    requested: string;
    expected: boolean;
  }>;
  scope_match_errors: Array<{
    name: string;
    granted: string[];
    requested: string;
    expect_error: string;
  }>;
  valid_tokens: Array<{
    name: string;
    description?: string;
    token: string;
    algorithm: "HS256" | "RS256";
    issuer: string;
    now: number;
    execution_id?: string | null;
    expected: {
      sub: string;
      jti: string;
      scopes: string[];
      execution_id: string | null;
      metadata: Record<string, unknown>;
    };
  }>;
  invalid_tokens: Array<{
    name: string;
    description?: string;
    token: string;
    algorithm: "HS256" | "RS256";
    issuer: string;
    now: number;
    execution_id?: string | null;
    expect_error: "expired" | "invalid" | "revoked";
  }>;
  issue_and_verify_roundtrip: Array<{
    name: string;
    description?: string;
    issue: {
      subject: string;
      scopes: string[];
      ttl_seconds: number;
      execution_id: string | null;
      metadata: Record<string, unknown> | null;
    };
    verify_with_execution_id?: string;
    expected_after_issue: {
      subject: string;
      scopes: string[];
      execution_id: string | null;
    };
    expected_after_verify_succeeds: boolean;
  }>;
}

const errorMap = {
  expired: ExpiredToken,
  invalid: InvalidToken,
  revoked: RevokedToken,
} as const;

export function runConsentVectors(specRoot: string): VectorResult[] {
  const data = JSON.parse(
    readFileSync(join(specRoot, "consent", "vectors.json"), "utf8"),
  ) as ConsentVectorFile;
  const sharedKey = data.hs256_key_shared;
  const results: VectorResult[] = [];

  // 1. scope_match
  for (const vec of data.scope_match ?? []) {
    const got = scopeGranted(vec.granted, vec.requested);
    if (got === vec.expected) {
      results.push({
        section: "consent",
        name: `scope_match/${vec.name}`,
        passed: true,
        message: "",
      });
    } else {
      results.push({
        section: "consent",
        name: `scope_match/${vec.name}`,
        passed: false,
        message: `expected ${JSON.stringify(vec.expected)}, got ${JSON.stringify(got)}`,
      });
    }
  }

  // 2. scope_match_errors
  for (const vec of data.scope_match_errors ?? []) {
    try {
      scopeGranted(vec.granted, vec.requested);
      results.push({
        section: "consent",
        name: `scope_match_error/${vec.name}`,
        passed: false,
        message: "expected error but call succeeded",
      });
    } catch (exc) {
      const msg = (exc as Error).message;
      if (msg.includes(vec.expect_error)) {
        results.push({
          section: "consent",
          name: `scope_match_error/${vec.name}`,
          passed: true,
          message: "",
        });
      } else {
        results.push({
          section: "consent",
          name: `scope_match_error/${vec.name}`,
          passed: false,
          message: `got error ${JSON.stringify(msg)} but expected substring ${JSON.stringify(vec.expect_error)}`,
        });
      }
    }
  }

  // 3. valid_tokens
  for (const vec of data.valid_tokens ?? []) {
    const verifier = new ConsentVerifier({
      verification_key: sharedKey,
      algorithm: vec.algorithm,
      issuer: vec.issuer,
      leeway_seconds: 0,
      clock: () => vec.now,
    });
    try {
      const scope = verifier.verify(vec.token, {
        execution_id: vec.execution_id ?? null,
      });
      const e = vec.expected;
      const mismatches: string[] = [];
      if (scope.subject !== e.sub)
        mismatches.push(`subject ${JSON.stringify(scope.subject)} != ${JSON.stringify(e.sub)}`);
      if (scope.jti !== e.jti)
        mismatches.push(`jti ${JSON.stringify(scope.jti)} != ${JSON.stringify(e.jti)}`);
      if (JSON.stringify(scope.scopes) !== JSON.stringify(e.scopes))
        mismatches.push(`scopes ${JSON.stringify(scope.scopes)} != ${JSON.stringify(e.scopes)}`);
      if (scope.execution_id !== e.execution_id)
        mismatches.push(`execution_id ${JSON.stringify(scope.execution_id)} != ${JSON.stringify(e.execution_id)}`);
      if (JSON.stringify(scope.metadata) !== JSON.stringify(e.metadata))
        mismatches.push(`metadata ${JSON.stringify(scope.metadata)} != ${JSON.stringify(e.metadata)}`);

      if (mismatches.length === 0) {
        results.push({
          section: "consent",
          name: `valid_token/${vec.name}`,
          passed: true,
          message: "",
        });
      } else {
        results.push({
          section: "consent",
          name: `valid_token/${vec.name}`,
          passed: false,
          message: mismatches.join("; "),
        });
      }
    } catch (exc) {
      results.push({
        section: "consent",
        name: `valid_token/${vec.name}`,
        passed: false,
        message: `verification failed: ${(exc as Error).name}: ${(exc as Error).message}`,
      });
    }
  }

  // 4. invalid_tokens
  for (const vec of data.invalid_tokens ?? []) {
    const verifier = new ConsentVerifier({
      verification_key: sharedKey,
      algorithm: vec.algorithm,
      issuer: vec.issuer,
      leeway_seconds: 0,
      clock: () => vec.now,
    });
    const expectedCls = errorMap[vec.expect_error];
    try {
      verifier.verify(vec.token, { execution_id: vec.execution_id ?? null });
      results.push({
        section: "consent",
        name: `invalid_token/${vec.name}`,
        passed: false,
        message: `expected ${expectedCls.name} but verify succeeded`,
      });
    } catch (exc) {
      if (exc instanceof expectedCls) {
        results.push({
          section: "consent",
          name: `invalid_token/${vec.name}`,
          passed: true,
          message: "",
        });
      } else {
        results.push({
          section: "consent",
          name: `invalid_token/${vec.name}`,
          passed: false,
          message: `expected ${expectedCls.name} but got ${(exc as Error).name}: ${(exc as Error).message}`,
        });
      }
    }
  }

  // 5. issue_and_verify_roundtrip
  for (const vec of data.issue_and_verify_roundtrip ?? []) {
    const issuer = new ConsentIssuer({
      signing_key: sharedKey,
      algorithm: "HS256",
      issuer: "iaiso",
    });

    try {
      const scope = issuer.issue({
        subject: vec.issue.subject,
        scopes: vec.issue.scopes,
        execution_id: vec.issue.execution_id,
        ttl_seconds: vec.issue.ttl_seconds,
        metadata: vec.issue.metadata ?? undefined,
      });
      const exp = vec.expected_after_issue;
      const mismatches: string[] = [];
      if (scope.subject !== exp.subject) mismatches.push("subject mismatch");
      if (JSON.stringify(scope.scopes) !== JSON.stringify(exp.scopes))
        mismatches.push("scopes mismatch");
      if (scope.execution_id !== exp.execution_id)
        mismatches.push("execution_id mismatch");

      if (mismatches.length > 0) {
        results.push({
          section: "consent",
          name: `roundtrip/${vec.name}`,
          passed: false,
          message: mismatches.join("; "),
        });
        continue;
      }

      const verifier = new ConsentVerifier({
        verification_key: sharedKey,
        algorithm: "HS256",
        issuer: "iaiso",
        leeway_seconds: 5,
      });
      verifier.verify(scope.token, {
        execution_id: vec.verify_with_execution_id ?? null,
      });
      results.push({
        section: "consent",
        name: `roundtrip/${vec.name}`,
        passed: true,
        message: "",
      });
    } catch (exc) {
      results.push({
        section: "consent",
        name: `roundtrip/${vec.name}`,
        passed: false,
        message: `roundtrip failed: ${(exc as Error).name}: ${(exc as Error).message}`,
      });
    }
  }

  return results;
}

File: IAIso-v5.0/core/iaiso-python/iaiso/conformance/_consent.py
"""Consent vector runner. Called from iaiso.conformance.run_consent_vectors."""

from __future__ import annotations

import time
from pathlib import Path

from iaiso.conformance import VectorResult, load_vectors


def run_consent_vectors(spec_root: Path) -> list[VectorResult]:
    from iaiso.consent import (
        ConsentIssuer,
        ConsentVerifier,
        ExpiredToken,
        InvalidToken,
        RevokedToken,
        _scope_granted,
    )

    data = load_vectors(spec_root, "consent")
    shared_key = data["hs256_key_shared"]
    results: list[VectorResult] = []

    # 1. Scope match vectors
    for vec in data.get("scope_match", []):
        got = _scope_granted(vec["granted"], vec["requested"])
        if got == vec["expected"]:
            results.append(VectorResult("consent", f"scope_match/{vec['name']}", True))
        else:
            results.append(VectorResult(
                "consent", f"scope_match/{vec['name']}", False,
                message=f"expected {vec['expected']!r}, got {got!r}",
            ))

    # 2. Scope match errors
    for vec in data.get("scope_match_errors", []):
        try:
            _scope_granted(vec["granted"], vec["requested"])
            results.append(VectorResult(
                "consent", f"scope_match_error/{vec['name']}", False,
                message=f"expected error but call succeeded",
            ))
        except ValueError as exc:
            if vec["expect_error"] in str(exc):
                results.append(VectorResult("consent", f"scope_match_error/{vec['name']}", True))
            else:
                results.append(VectorResult(
                    "consent", f"scope_match_error/{vec['name']}", False,
                    message=f"got error {exc!s} but expected substring {vec['expect_error']!r}",
                ))

    # 3. Valid tokens — must verify cleanly with given key/now/execution
    for vec in data.get("valid_tokens", []):
        verifier = ConsentVerifier(
            verification_key=shared_key,
            algorithm=vec["algorithm"],
            issuer=vec["issuer"],
            leeway_seconds=0,
        )
        try:
            _with_frozen_time(vec["now"], lambda: verifier.verify(
                vec["token"], execution_id=vec.get("execution_id"),
            ))
        except Exception as exc:  # noqa: BLE001
            results.append(VectorResult(
                "consent", f"valid_token/{vec['name']}", False,
                message=f"verification failed: {type(exc).__name__}: {exc}",
            ))
            continue

        # Re-verify to check expected fields
        scope = _with_frozen_time(vec["now"], lambda: verifier.verify(
            vec["token"], execution_id=vec.get("execution_id"),
        ))
        expected = vec["expected"]
        mismatches = []
        if scope.subject != expected["sub"]:
            mismatches.append(f"subject {scope.subject!r} != {expected['sub']!r}")
        if scope.jti != expected["jti"]:
            mismatches.append(f"jti {scope.jti!r} != {expected['jti']!r}")
        if scope.scopes != expected["scopes"]:
            mismatches.append(f"scopes {scope.scopes!r} != {expected['scopes']!r}")
        if scope.execution_id != expected["execution_id"]:
            mismatches.append(f"execution_id {scope.execution_id!r} != {expected['execution_id']!r}")
        if dict(scope.metadata) != expected["metadata"]:
            mismatches.append(f"metadata {scope.metadata!r} != {expected['metadata']!r}")

        if mismatches:
            results.append(VectorResult(
                "consent", f"valid_token/{vec['name']}", False,
                message="; ".join(mismatches),
            ))
        else:
            results.append(VectorResult("consent", f"valid_token/{vec['name']}", True))

    # 4. Invalid tokens — must raise the expected error kind
    error_map = {
        "expired": ExpiredToken,
        "invalid": InvalidToken,
        "revoked": RevokedToken,
    }
    for vec in data.get("invalid_tokens", []):
        verifier = ConsentVerifier(
            verification_key=shared_key,
            algorithm=vec["algorithm"],
            issuer=vec["issuer"],
            leeway_seconds=0,
        )
        expected_cls = error_map[vec["expect_error"]]
        try:
            _with_frozen_time(vec["now"], lambda: verifier.verify(
                vec["token"], execution_id=vec.get("execution_id"),
            ))
            results.append(VectorResult(
                "consent", f"invalid_token/{vec['name']}", False,
                message=f"expected {expected_cls.__name__} but verify succeeded",
            ))
        except expected_cls:
            results.append(VectorResult("consent", f"invalid_token/{vec['name']}", True))
        except Exception as exc:  # noqa: BLE001
            results.append(VectorResult(
                "consent", f"invalid_token/{vec['name']}", False,
                message=f"expected {expected_cls.__name__} but got {type(exc).__name__}: {exc}",
            ))

    # 5. Issue-and-verify roundtrip
    for vec in data.get("issue_and_verify_roundtrip", []):
        issuer = ConsentIssuer(signing_key=shared_key, algorithm="HS256", issuer="iaiso")
        issue_args = vec["issue"]
        try:
            scope = issuer.issue(
                subject=issue_args["subject"],
                scopes=issue_args["scopes"],
                execution_id=issue_args.get("execution_id"),
                ttl_seconds=issue_args.get("ttl_seconds"),
                metadata=issue_args.get("metadata"),
            )
        except Exception as exc:  # noqa: BLE001
            results.append(VectorResult(
                "consent", f"roundtrip/{vec['name']}", False,
                message=f"issue failed: {type(exc).__name__}: {exc}",
            ))
            continue

        # Check issued fields
        exp = vec["expected_after_issue"]
        mismatches = []
        if scope.subject != exp["subject"]:
            mismatches.append(f"subject mismatch")
        if scope.scopes != exp["scopes"]:
            mismatches.append(f"scopes mismatch")
        if scope.execution_id != exp["execution_id"]:
            mismatches.append(f"execution_id mismatch")

        if mismatches:
            results.append(VectorResult(
                "consent", f"roundtrip/{vec['name']}", False,
                message="; ".join(mismatches),
            ))
            continue

        # Verify
        verifier = ConsentVerifier(
            verification_key=shared_key, algorithm="HS256", issuer="iaiso", leeway_seconds=5,
        )
        try:
            verifier.verify(
                scope.token,
                execution_id=vec.get("verify_with_execution_id"),
            )
            results.append(VectorResult("consent", f"roundtrip/{vec['name']}", True))
        except Exception as exc:  # noqa: BLE001
            results.append(VectorResult(
                "consent", f"roundtrip/{vec['name']}", False,
                message=f"verify after issue failed: {type(exc).__name__}: {exc}",
            ))

    return results


def _with_frozen_time(frozen_now: float, fn):
    """Call `fn` with time.time() returning `frozen_now`. Restores after."""
    import iaiso.consent
    orig = iaiso.consent.time.time
    iaiso.consent.time.time = lambda: frozen_now
    try:
        # PyJWT reads time directly via datetime.now/utcnow internally — we
        # need to also patch datetime. Easier: patch PyJWT's "now".
        import jwt.api_jwt
        orig_now = None
        if hasattr(jwt.api_jwt, 'datetime'):
            import datetime as _dt
            orig_dt = jwt.api_jwt.datetime
            class _Frozen(_dt.datetime):
                @classmethod
                def now(cls, tz=None):
                    return _dt.datetime.fromtimestamp(frozen_now, tz=tz or _dt.timezone.utc)
                @classmethod
                def utcnow(cls):
                    return _dt.datetime.utcfromtimestamp(frozen_now)
            jwt.api_jwt.datetime = _Frozen
            try:
                return fn()
            finally:
                jwt.api_jwt.datetime = orig_dt
        else:
            return fn()
    finally:
        iaiso.consent.time.time = orig
